spec/java/util/ArrayList.lsl

```diff
@@ -0,0 +1,1205 @@
+        action NOT_IMPLEMENTED();
+
+        if (modCount != expectedModCount)
+        {
+            action THROW_NEW("java.util.ConcurrentModificationException", []);
+        }
+
+        modCount = modCount + 1;
+    }
+}
+
+
+
+@packagePrivate
+@implements("java.util.Iterator")
+@WrapperMeta(
+    src="java.util.ArrayList$Itr",
+    //Maybe will be another name of the dst class
+    dst="ru.spbpu.libsl.overrides.collections.ArrayList_Itr",
+    forceMatchInterfaces=true)
+automaton Itr: int (
+    var cursor: int,
+    var lastRet: int = -1,
+    var expectedModCount: int
+) {
+
+    initstate Initialized;
+
+    shift Allocated -> Initialized by [
+        Itr()
+    ];
+
+    shift Initialized -> self by [
+        // read operations
+        hasNext,
+
+        // write operations
+        next,
+        remove,
+        forEachRemaining
+    ]
+
+
+    // constructors
+
+    constructor Itr()
+    {
+    }
+
+
+    // methods
+
+
+    fun hasNext (): boolean
+    {
+        result = cursor != self.parent.length;
+    }
+
+
+    fun next (): Object
+    {
+        self.parent._checkForComodification(expectedModCount);
+        var i = cursor;
+
+        if (i >= parent.length)
+        {
+            action THROW_NEW("java.util.NoSuchElementException", []);
+        }
+
+        // #problem
+        //I don't know what to do with ConcurrentModificationException(); ?
+
+        cursor = i + 1;
+        lastRet = i;
+        result = action LIST_GET(self.parent.storage, lastRet);
+    }
+
+
+    fun remove (): void
+    {
+        if (lastRet < 0)
+        {
+            action THROW_NEW("java.lang.IllegalStateException", []);
+        }
+
+        self.parent._checkForComodification(expectedModCount);
+
+        // #problem
+        //What i must to do with try-catch in this method ?
+
+        self.parent._deleteElement(lastRet);
+        cursor = lastRet;
+        lastRet = -1;
+        expectedModCount = self.parent.modCount;
+    }
+
+
+    fun forEachRemaining (action: Consumer): void
+    {
+        action NOT_IMPLEMENTED();
+    }
+
+}
+
+
+
+@packagePrivate
+@extends("java.util.ArrayList$Itr")
+@implements("java.util.Iterator")
+@WrapperMeta(
+    src="java.util.ArrayList$ListItr",
+    //Maybe will be another name of the dst class
+    dst="ru.spbpu.libsl.overrides.collections.ArrayList_ListItr",
+    forceMatchInterfaces=true)
+automaton ListItr: int (
+    var cursor: int,
+    var lastRet: int = -1,
+    var expectedModCount: int
+) {
+
+    initstate Initialized;
+
+    shift Allocated -> Initialized by [
+        ListItr(int)
+    ];
+
+    shift Initialized -> self by [
+       // read operations
+       hasNext,
+       hasPrevious,
+       nextIndex,
+       previousIndex,
+
+       // write operations
+       next,
+       remove,
+       forEachRemaining,
+       previous,
+       set,
+       add
+    ]
+
+    //constructors
+
+
+    constructor ListItr(index: int)
```

Comment:
It is not possible (except reflection) to call the constructors of this class outside of the `ArrayList`.

Reply:
Fixed